File: src/p0002/solution.go
/**
 * Definition for singly-linked list.
 * type ListNode struct {
 *     Val int
 *     Next *ListNode
 * }
 */
 func addTwoNumbers(l1 *ListNode, l2 *ListNode) *ListNode {
	carry := 0
	p1, p2 := l1, l2
	var ans, p *ListNode
	for p1 != nil || p2 != nil {
		if p == nil {
			ans = &ListNode{}
			p = ans
		} else {
			p.Next = &ListNode{}
			p = p.Next
		}
		if p1 != nil {
			p.Val += p1.Val
		}
		if p2 != nil {
			p.Val += p2.Val
		}
		p.Val += carry
		carry = p.Val / 10
		p.Val %= 10
		if p1 != nil {
			p1 = p1.Next
		}
		if p2 != nil {
			p2 = p2.Next
		}
	}
	if carry != 0 {
		p.Next = &ListNode{
			Val: 1,
		}
	}
	return ans
}

File: src/p0002/python/solution.py
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, val=0, next=None):
#         self.val = val
#         self.next = next
class Solution:
    def addTwoNumbers(self, l1: Optional[ListNode], l2: Optional[ListNode]) -> Optional[ListNode]:
        p1, p2 = l1, l2
        plus = 0
        dummy = ListNode()
        p = dummy
        while p1 or p2 or plus:
            p.next = ListNode()
            if p1:
                p.next.val += p1.val
                p1 = p1.next
            if p2:
                p.next.val += p2.val
                p2 = p2.next
            p.next.val += plus
            plus = p.next.val // 10
            p.next.val %= 10
            p = p.next
        
        return dummy.next